codec-http2/src/main/java/io/netty/handler/codec/http2/Http2ConnectionHandler.java

```diff
@@ -378,16 +384,34 @@ protected Http2FrameWriter frameWriter() {
      * Writes a {@code RST_STREAM} frame to the remote endpoint and updates the connection state appropriately.
      */
     @Override
-    public ChannelFuture writeRstStream(ChannelHandlerContext ctx, int streamId, long errorCode,
-            ChannelPromise promise) {
-        Http2Stream stream = connection().stream(streamId);
+    public ChannelFuture writeRstStream(final ChannelHandlerContext ctx, int streamId, long errorCode,
+            final ChannelPromise promise) {
+        final Http2Stream stream = connection().stream(streamId);
+        if (stream == null) {
+            // Don't write a RST_STREAM frame if we are not aware of the stream.
+            return promise.setFailure(streamError(streamId, INTERNAL_ERROR,
+                    "Stream %d not found while writing RST_STREAM frame", streamId));
+        }
+        if (stream.isResetSent()) {
+            // Don't write a RST_STREAM frame if we already wrote a RST_STREAM frame
+            return promise.setFailure(streamError(streamId, INTERNAL_ERROR,
```

Comment:
> We have this in GRPC actually where close could be called multiple times on different ends and thus it would lead to several calls to writeRstStream.

Can you enlighten me?  Seems counter intuitive to call a method which returns "success" when an event (sending something on the wire) happens as a result of that call.  Considering state seems like it would make it *more* difficult for debuggers to track down issues.  I just envision trying to debug a problem and getting a success back from the interface...assuming that means a RST was sent...but not actually being a RST go out on the wire.  For example we could have been in trouble (and may still be in trouble elsewhere) if we synchronously mark the event as occurred but a failure happens asynchronously (i.e. the write operation fails).  Then we will mark the state as "the event happened" which will in the future prevent the event from ever really happening.